### PR TITLE
add a way to plumb through different account providers

### PR DESCRIPTION
**Pull request by kahing**

### CODE REVIEW

**Comment by andrewgaul** [comment]
Can you expand on the commit message?  I don't understand how this is supposed to be used, especially the interaction with the default `BlobStore` and authentication.  Instead should we remove the notion of a default?

File: src/main/java/org/gaul/s3proxy/S3ProxyHandler.java [hunk 1]
```diff
@@ -1610,6 +1655,11 @@ private void sendSimpleErrorResponse(HttpServletResponse response,
         }
     }
 
+    public void setProviders(
+            Map<String, Map.Entry<String, BlobStore>> providers) {
+        this.providers = providers;
```
Comment (andrewgaul) [inline src/main/java/org/gaul/s3proxy/S3ProxyHandler.java]: Should this defensively copy, or instead just offer an add and remove interface?

File: src/main/java/org/gaul/s3proxy/S3ProxyHandler.java [hunk 2]
```diff
@@ -1464,9 +1509,9 @@ private void handleListParts(HttpServletRequest request,
         }
     }
 
-    private void handleUploadPart(HttpServletRequest request,
-            HttpServletResponse response, String containerName,
-            String blobName, String uploadId)
+    private void handleUploadPart(BlobStore blobStore,
+            HttpServletRequest request, HttpServletResponse response,
+            String containerName, String blobName, String uploadId)
```
Comment (andrewgaul) [inline src/main/java/org/gaul/s3proxy/S3ProxyHandler.java]: Can you reorder these parameters?  `blobStore` logically belongs next to `containerName`.

File: src/main/java/org/gaul/s3proxy/S3ProxyHandler.java [hunk 3]
```diff
@@ -316,79 +332,84 @@ private void doHandle(String target, Request baseRequest,
...
-                    return;
-                }
-                handleBlobList(request, response, path[1]);
+                handleContainerList(blobStore, response);
                 baseRequest.setHandled(true);
                 return;
             } else {
-                if ("".equals(request.getParameter("acl"))) {
-                    handleGetBlobAcl(response, path[1], path[2]);
+                if (path.length <= 2 || path[2].isEmpty()) {
```
Comment (andrewgaul) [inline src/main/java/org/gaul/s3proxy/S3ProxyHandler.java]: Why code movement?

File: src/main/java/org/gaul/s3proxy/S3ProxyHandler.java [hunk 4]
```diff
@@ -161,16 +162,26 @@
...
+            defaultBlobStore = blobStore;
+        }
         this.virtualHost = requireNonNull(virtualHost);
         xmlOutputFactory.setProperty("javax.xml.stream.isRepairingNamespaces",
                 Boolean.FALSE);
     }
 
+    private <K, V> Map.Entry<K, V> makePair(K first, V second) {
+        return new AbstractMap.SimpleImmutableEntry<>(first, second);
+    }
```
Comment (andrewgaul) [inline src/main/java/org/gaul/s3proxy/S3ProxyHandler.java]: Replace helper with Guava `Maps.immutableEntry`.

File: src/main/java/org/gaul/s3proxy/S3ProxyHandler.java [hunk 5]
```diff
@@ -247,54 +258,59 @@ private void doHandle(String target, Request baseRequest,
...
-            String headerSignature = null;
-            if (headerAuthorization != null &&
-                    headerAuthorization.startsWith("AWS ")) {
+        BlobStore blobStore;
+        String requestIdentity = null;
+        String requestSignature = null;
+        String headerAuthorization = request.getHeader(
+                HttpHeaders.AUTHORIZATION);
+
+        if (headerAuthorization != null) {
```
Comment (andrewgaul) [inline src/main/java/org/gaul/s3proxy/S3ProxyHandler.java]: Could we refactor authentication into a separate commit?

Comment (kahing) [inline src/main/java/org/gaul/s3proxy/S3ProxyHandler.java]: no, the way it was done makes it hard to do the lookup

File: src/main/java/org/gaul/s3proxy/S3ProxyHandler.java (hunk 4 above)
Comment (kahing) [inline src/main/java/org/gaul/s3proxy/S3ProxyHandler.java]: fixed

File: src/main/java/org/gaul/s3proxy/S3ProxyHandler.java (hunk 3 above)
Comment (kahing) [inline src/main/java/org/gaul/s3proxy/S3ProxyHandler.java]: fixed

File: src/main/java/org/gaul/s3proxy/S3ProxyHandler.java (hunk 2 above)
Comment (kahing) [inline src/main/java/org/gaul/s3proxy/S3ProxyHandler.java]: fixed

File: src/main/java/org/gaul/s3proxy/S3ProxyHandler.java (hunk 1 above)
Comment (kahing) [inline src/main/java/org/gaul/s3proxy/S3ProxyHandler.java]: fixed

**Comment by andrewgaul** [comment]
I made some minor changes and pushed to master as 
commit 87bed283cbe39f0b1ba7bf5ec1c749dba2b21d0f.

**Pull request closed without merging**